docs: add blog about the power of secret store (#148)
arcus-azure/arcus: community/blog/2021-07 The power of Arcus secret store/README.md

## What is it again?

The Arcus secret store is a system that provides secrets located in different places via a single access point. In a more complex application, secrets values can be located in Azure Key Vaults, environment variables, development configuration files... The Arcus secret store lets you access all of them with a single call instead of using specific SDK's to access them separatly.
In many ways, the secret store looks simulair to the [ASP.NET Core configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/) with some subtle but powerful differences.

Some terminology:
* **Secret**: sensitive information that needs to be stored securely. Access keys, connection strings, client secrets...
* **Secret provider**: instance that provides secrets from a certain location. This location can be: Azure Key Vault, environment variables, HashiCorp Vault... and your own locations.
* **Secret store**: system that combines all the secret providers into a single access point.

> An example of a simple secret store configuration is shown here:

Author: Great stuff!

                {
                    config.AddJsonFile("appsettings.json")
                          .AddJsonFile("appsettings.Development.json");
                })
                .ConfigureSecretStore((context, config, builder) =>
                {
#if DEBUG
                    builder.AddConfiguration(config);
#endif
                    var keyVaultName = config["KeyVault_Name"];
                    builder.AddEnvironmentVariables()
                           .AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net");
                });
    }
} 
```

> Note that when a secret is being retrieved, the secret providers are being called in the order they were registered. So in this case, the environment variables will be checked before the Azure Key Vault is contacted.

The secret store is also available for [Azure Functions](https://security.arcus-azure.net/features/secret-store/#using-secret-store-within-azure-functions) and [can be configured outside the `Program`](https://security.arcus-azure.net/features/secret-store/#configuring-secret-store-without-net-host-builder).
Let's look at the advantages of using our Arcus secret store.

## Caching

> A feature that makes the secret store unique, is caching. This makes sure that registered secret providers won't be called upon every secret retrieval. This helps you avoiding hitting service limitations.
> The secret store also allows you to define your own caching strategy for [custom secret providers](https://security.arcus-azure.net/features/secret-store/create-new-secret-provider) or extending existing ones.
> 
> The caching functionality acts as a 'wrapper' arround existing secret providers. Any secret provider you can think of can be 'wrapped' into a cached version.

Reviewer: ```suggestion
A feature that makes the secret store unique, is caching. This allows the secret providers to not hit service limitations based on your usage pattern and improve performance.

The secret store also allows you to define your own caching strategy for [custom secret providers](https://security.arcus-azure.net/features/secret-store/create-new-secret-provider) or extending existing ones.

The caching functionality acts as a 'wrapper' around existing secret providers and thus any secret provider you can think of can be 'wrapped' into a cached version.
```

```csharp
using Arcus.Security.Core.Caching;
using Arcus.Security.Core.Caching.Configuration;
using Arcus.Security.Core.Providers;

.ConfigureSecretStore((config, stores) =>
{
    // Configuration cache (optional: default is 5 min).
    var cacheConfiguration = new CacheConfiguration(TimeSpan.FromMinutes(10));

    // Built-in secret providers.
    var environmentVariables = new EnvironmentVariableSecretProvider();
    var cachedEnvironmentVariables = new CachedSecretProvider(environmentVariables, cacheConfiguration);

    // Custom secret providers.
    var windowsRegistry = new WindowsRegistrySecretProvider();
    var cachedWindowsRegistry = new CahcedSecretProvider(windowsRegistry, CacheConfiguration.Default);
});
```

> See also our [asynchrounous cache invalidation](https://background-jobs.arcus-azure.net/features/security/auto-invalidate-secrets) to automatically invalidate cached Azure Key Vault secrets. 

Reviewer: ```suggestion
We allow you to [asynchrounously invalidate the cache](https://background-jobs.arcus-azure.net/features/security/auto-invalidate-secrets) to automatically keep secrets aligned with the latest Azure Key Vault secrets. 
```

## Plug & play

> Our secret store is very easily to set up. All it requires is a single call to register the store `ConfigureSecretStore`, after that you can immediately use our `ISecretProvider` interface abstraction throughout your application.

Reviewer: ```suggestion
Our secret store is very easy to set up. All that is required is defining what your secret store looks like through `ConfigureSecretStore`, after that you can immediately use our `ISecretProvider` interface abstraction throughout your application.
```

```csharp
using Arcus.Security.Core;

[ApiController]
public class OrderController : ControllerBase
{
    private readonly ISecretProvider _secretStore;

    public OrderController(ISecretProvider secretStore)
    {
        _secretStore = secretStore;
    }

    [HttpGet]
    public async Task<IActionResult> Orders()
    {
        string secret = await _secretStore.GetRawSecretAsync("MySecret");
    }
}
```

> We are very flexible in the registration of the secret providers. We allow all kinds of combinations and even multiple registrations of the same type. Imagine two Azure Key Vaults with public and user secrets.

Reviewer: ```suggestion
We are very flexible in the registration of the secret providers allowing all kinds of combinations and even multiple registrations of the same type, for example, you can use two Azure Key Vaults sources.
```

```csharp
using Microsoft.Extensions.Hosting;

.ConfigureSecretStore((config, stores) =>
{

>     stores.AddAzureKeyVaultWithManagedServiceIdentity("https://public-secrets.vault.azure.net")

Reviewer: ```suggestion
    stores.AddAzureKeyVaultWithManagedServiceIdentity("https://shared-platform.vault.azure.net")
```

>           .AddAzureKeyVaultWithManagedServiceIdentity("https://user-secrets.vault.azure.net"); 

Reviewer: ```suggestion
          .AddAzureKeyVaultWithManagedServiceIdentity("https://orders.vault.azure.net"); 
```

})
```

> Note that with our [named secret providers feature](https://security.arcus-azure.net/features/secret-store/named-secret-providers), you can retrieve a single secret provider instead of the combination. This can be useful in a scenario like this.

> ## Designed for security

Reviewer: Let's move this to the top, this is the trigger for building the secret store

Author: Done that. Thanks! I agree.

Besides all the user-friendly features, there're also security reasons why the Arcus secret store is a recommended choice. We made sure that all the secret providers are registered in one single place and are being contacted via one single access point. This is much safer and easier to maintain then writing multiple service integrations yourself, scattered across the application. 

> It's also safer than using the general [ASP.NET Core configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/) approach because we explicitly made the distinction between **data** and **secrets**. Many vulnerabilities gets introduced when secrets are seeen as data and are included in logs, for example. Leakage of sensitive information is a real risk in this case. Or when expired secrets doesn't get transient handling upon retrieval.

Reviewer: Let's move this as the first argument, this is essential

Author: Aha, your suggestion already made this alteration. Thanks!

> Besides all the user-friendly features, there're also security reasons why the Arcus secret store is a recommended choice. We made sure that all the secret providers are registered in one single place and are being contacted via one single access point. This is much safer and easier to maintain then writing multiple service integrations yourself, scattered across the application. 
> 
> It's also safer than using the general [ASP.NET Core configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/) approach because we explicitly made the distinction between **data** and **secrets**. Many vulnerabilities gets introduced when secrets are seeen as data and are included in logs, for example. Leakage of sensitive information is a real risk in this case. Or when expired secrets doesn't get transient handling upon retrieval.

Reviewer: ```suggestion

While [.NET's configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/) approach feels very similar, it does not make an explicit distinction between **configuration** and **secrets**.

Many vulnerabilities get introduced when secrets are seen as data and are included in logs, for example. Leakage of sensitive information is a real risk in this case. Or when expired secrets don't get transient handling upon retrieval.

**With Arcus secret store, you are making secrets a first-class citizen in .NET Core** to ensure they are handled accordingly.

We make sure that all the secret providers are registered in a central place and are being contacted via the secret store. This is much safer and easier to maintain since it is easy to get an overview of the various secret sources that the application depends on.

Lastly, there is a lot less work for end-users since they no longer have to write multiple service integrations themselves, scattered across the application, but instead rely on a variety of sources.
```

> We also included [security events](https://security.arcus-azure.net/features/secret-store/#include-security-auditing) to our secret store to make sure that malicous activity can be detected more easily. These events are written every time the secret store is contacted.

Reviewer: ```suggestion
We also provide [security events](https://security.arcus-azure.net/features/secret-store/#include-security-auditing) to our secret store to make sure that malicious activity can be detected more easily. These events are written every time the secret store is queried.
```

> With all the scecret-contacting functionality in one place, you're not only making your application more secure but decreasing the maintenance effort.

Reviewer: ```suggestion
With all the secret functionality in one place, you're not only making your application more secure but decreasing the maintenance effort.
```

## Extensibility

> A last section we can dedicate to the extensibility of the Arcus secret store. You can write your own secret providers, use [in-memory secret providers](https://github.com/arcus-azure/arcus.testing/blob/master/docs/features/inmemory-secret-provider.md) for testing purposes. We are also working hard to make sure that *all* built-in secret providers are completely overridable so you can bring your own version of the existing functionality. If that's not enough, you can also [mutate secret names](https://security.arcus-azure.net/features/secret-store/create-new-secret-provider#adding-secret-name-mutation-before-looking-up-secret) on-the-fly when they come into any secret provider.
> 
> Even the registration process of the secret store is adaptable. We have demonstrated this [here](https://www.codit.eu/blog/role-based-authorization-low-level-customization-arcus-secret-store/) where we introduced an extra layer of authorization that determined which user role gets to access which secret provider.

Reviewer: ```suggestion
We are open to extension!

As an example, we allow you to write your own secret providers to retrieve secrets where you need them or you can use [in-memory secret providers](https://github.com/arcus-azure/arcus.testing/blob/master/docs/features/inmemory-secret-provider.md) for testing purposes.

We are also working hard to make sure that *all* built-in secret providers are completely extensible so you can bring your own version of the existing functionality. If that's not enough, you can also [mutate secret names](https://security.arcus-azure.net/features/secret-store/create-new-secret-provider#adding-secret-name-mutation-before-looking-up-secret) on-the-fly when they come into any secret provider.

Even the registration process of the secret store is adaptable. We have demonstrated this [here](https://www.codit.eu/blog/role-based-authorization-low-level-customization-arcus-secret-store/) where we introduced an extra layer of authorization that determined which user role gets to access which secret provider.
```
